### ISSUE TITLE
Layout / Positioning API Draft

### ISSUE THREAD

**Issue by mitchmindtree** [PistonDevelopers/conrod#380]
Discussion started [here](https://github.com/PistonDevelopers/conrod/issues/374#issuecomment-93914098).

Summary
-------------
Conrod should require that a user **never** has to think about or use the coordinate system. If a use case is found in which the user finds they have no choice but to use coordinates directly, then we should treat this as a serious API design flaw and address it.

To achieve this, we need an expressive and thorough API. I propose the following design.

Canvases (i.e. splitting up the window into sections)
-------------------------------------------------------------------
This should be the first step in the user's approach to widget layout.

In almost all cases, you want to be able to divide the screen into different sections (i.e. columns and rows) and have a custom name you can use to describe each section (i.e. `header`, `mid_left_column`).

To do this, I propose using something like @eddyb's super simple `vbox`/`hbox` (short for "vertical box" and "horizontal box") design.

Here are some pics showing eddyb's implementation [1](https://i.imgur.com/btcNQsV.png), [2](https://i.imgur.com/btbVGSw.png) and [3](https://i.imgur.com/WfJMqDW.png).

We could provide `vbox!` and `hbox!` macros to make this simple and easy.

For example:

```Rust
vbox![header, hbox![left_col, mid_col, right_col], footer];
```

would translate to 5 canvases:
- A `Canvas` called `header` occupying the top third of the screen.
- Three column `Canvas`es occuping the centre third of the screen named `left_col`, `mid_col` and `right_col`.
- A `Canvas` called `footer` occupying the bottom third of the screen.

The `Ui`s active `Theme` should specify default margins and padding, however we should also allow this to be described in the `vbox!` and `hbox!` macros, i.e.:

```Rust
vbox![
    header: top_padding = 20,
    hbox![left_col: left_margin = 30; top_padding = 30,
          mid_col: left_margin = 30; top_padding = 30,
          right_col: left_margin = 30; top_padding = 30],
    footer: bottom_padding = 50,
];
```

So far we're able to split up a screen into even sections, however a user may sometimes want to specify non-even splits with specific percentages of the screen. We could also describe this within our `vbox!` and `hbox!` macros:

```Rust
vbox![
    header: split = 20%,
    hbox![left_col: split = 50%, mid_col, right_col]: split = 60%,
    footer,
];
```

This would would result in:
- the `header` `Canvas` occupying the top 20% of the window.
- the three column `Canvas`es occupying the next 60% (vertically) with the `left_col` occupying the first horizontal 50% and the other two columns occupying 25% each.
- the `footer` `Canvas` occupying the remaining vertical percentage of the window (bottom 20%).

To draw a background for a specific canvas, we could add a `.canvas` method to the `Background` type, i.e.
```Rust
Background::new().canvas(&header).rgb(r, g, b)...
```

Widget placement within `Canvas`es
-------------------------------------------------
To place a widget within a canvas, we can provide a collection of helper methods:
```Rust
Positionable::{
    top_left, top_centre, top_right,
    middle_left, middle_centre, middle_right,
    bottom_left, bottom_centre, bottom_right
}
```
Which (using our canvases above) could be used like this:
```Rust
Label::new(TITLE, title).middle_centre(&header).etc().etc();
Slider::new(SLIDER, ...).top_left(&left_col).etc().ect();
```

This would resulting in drawing TITLE directly in the centre of the `header` canvas, and SLIDER in the top left corner of the `left_col` canvas.

Relative positioning of widgets
-----------------------------------------
From here on, all widget positioning could easily be described relatively using the following methods:
```Rust
Positionable::{
    left, right, up, down
    left_from, right_from, up_from, down_from
}
```
where `left`, `right`, `up`, `down` are relative to the last widget, and `left_from`, `right_from`, `down_from`, `up_from` are relative to some widget with the given `UiId`. Using our above widgets, these methods could be used like this:
```Rust
Label::new(SUBTITLE, subtitle).down_from(TITLE, 50.0).etc().etc();
Slider::new(SLIDER+1, ...).down_from(SLIDER, 20.0).etc().etc();
Slider::new(SLIDER+2, ...).down(20.0).etc().etc();
```

This would result in drawing a SUBTITLE 50 pixels below the TITLE, and two more sliders directly under the first SLIDER.

Thoughts
------------

Let me know any thoughts / suggestions on this! I'm happy to implement this as I'll be continuing to focus on conrod over the next couple days.

cc @bvssvni @dcampbell24 
related to #349, #350, #354, #374

**Comment by eddyb** [comment]
Just a reminder, [the current demo](https://github.com/eddyb/r3/blob/master/src/main.rs#L52) uses a single `flow!` macro instead of having a `vbox`/`hbox` distinction.

**Comment by bvssvni** [comment]
Being able to change the size by dragging the edge is also useful in some interfaces.

**Comment by bvssvni** [comment]
I don't like percentages for the reason that it seems too simple model to solve a difficult problem, and it needs to add up to 100% (I guess we usually are cheating and treating it as a weight). Often you have a sidebar which you want to adjust the edge manually with a minimum size, while the rest occupies all available area. It would be nice if you could specify a minimum and maximum and a weight for when it's between the minimum and maximum values. Weights are much nicer for adding and removing widgets.

**Comment by mitchmindtree** [comment]
@bvssvni hmm this is a good point, I was picturing the macros expanding to something like `const $canvas: Canvas = ...; (with the split percentages represented between 0.0 and 1.0), but it would be really nice to support adjustable canvases as well.

Perhaps instead they could expand into local bindings, and then the user can do something like `ui.register(canvas)` or `canvas.set(ui)` so that they can be cached within the `Ui` and react to input events (i.e. have their dimensions updated accordingly)?

**Comment by eddyb** [comment]
In my linear system of position variables, all you would have to do to would be to constrain, e.g. `x2 - x1` to some value (or a range), if you wanted to control its width.

Weights are also possible to support, and combined with per-variable "growth" directionality, each side of the bounding box could have individual behavior.

As an example relevant to my usecase, a menu or a toolbar would have "fill" behavior horizontally (x1 shrinks, x2 grows) while vertically it would act as a "shrinkwrap" (y1 and y2 both shrink).

**Comment by dcampbell24** [comment]
I wonder if we could take some inspiration from  [Grid-Flavored VFL](http://gridstylesheets.org/guides/vfl/). They seem to incorporate a lot of the ideas mentioned above, but the system is a more sophisticated (and complex). Fortunately, it also appears to be quite well documented and thought out. It is a system relating to CSS, so we would have to make some modifications, but we could probably use macros to mimic their VFL DSL for some rather ergonomic layout.

Thoughts?

**Comment by mitchmindtree** [comment]
Closed by #433, #410 and #385.

Still don't have a helper macro implemented but will add a separate issue for this.
